clash-lib/src/Clash/Backend/VHDL.hs

```diff
@@ -1509,7 +1558,12 @@ expr_ _ e@(DataCon ty@(Vector _ elTy) _ [e1,e2]) = do
   syn <- Ap hdlSyn
   case syn of
     Vivado -> qualTyName ty <> "'" <> case vectorChain e of
-      Just es -> align (tupled (mapM (toSLV elTy) es))
+      -- If the elements of the vector are an enumeration type, the additional
+      -- call to toSLV here results in the wrong code being generated, as the
+      -- values of the array _should be_ enum variants instead of SLVs.
+      Just es
+        | Sum{} <- elTy -> align (tupled (traverse (expr_ False) es))
```

Comment:
When we create VHDL for Vivado, the element of vectors should always be `std_logic_vector`. See https://github.com/clash-lang/clash-compiler/blob/c200bcc03ff7270e633c6d0070997bb82410b081/clash-lib/src/Clash/Backend/VHDL.hs#L481

Did you a testsuite run with `-fclash-hdlsyn Vivado`?

Reply:
Yes, `T1360` has `-fclash-hdlsyn Vivado` and was failing. Looking at it again, I can see that I've removed the wrong call (I _should_ have removed a `fromSLV` that occurs in the same area). I'm surprised that `ghdl` didn't complain about the type mismatch...

Reply:
Alright, on second (third?) look I've got a nicer solution. The problem is basically that with the sum types it would generate something like

```vhdl
x <= my_array_type'(toSLV(fromSLV(x_0)), ..., toSLV(fromSLV(x_n)));
```

which fails because it can't pick a sensible overload for `toSLV` / `fromSLV`. So when targeting Vivado array elements are now rendered by checking if the expression given to `toSLV` has `FromBv` as the outermost constructor, and removing both conversions if that is the case.

So for Vivado it now generates

```vhdl
x <= my_array_type'(x_0, ..., x_n);
```

and for non-vivado where array elements are the enum type it generates the usual

```vhdl
x <= my_array_type'(fromSLV(x_0), ..., fromSLV(x_n));
```